lsp/nls/src/requests/completion.rs

```diff
@@ -21,56 +21,85 @@ use crate::{
+                let new_id = fields.get(&name)?;
+                find_fields_from_term_kind(&linearization, *new_id, path)
+            }
+        }
+        TermKind::RecordField {
+            value: ValueState::Known(new_id),
+            ..
+        }
+        | TermKind::Declaration(_, _, ValueState::Known(new_id))
         | TermKind::Usage(UsageState::Resolved(new_id))
             if item.id == id =>
         {
-            find_record_fields(linearization, new_id)
+            find_fields_from_term_kind(linearization, new_id, path)
         }
         _ => None,
     }
 }
 
-/// Find a record contract of the item with the specified id.
-fn find_contract_record_fields(linearization: &Completed, id: usize) -> Option<Vec<Ident>> {
+/// Find the record fields associated with an ID in the linearization using
+/// its contract information.
+fn find_fields_from_contract(
+    linearization: &Completed,
+    id: usize,
+    path: &mut Vec<Ident>,
+) -> Option<Vec<Ident>> {
     let item = linearization.get_item(id)?;
     match &item.meta {
         Some(MetaValue { contracts, .. }) if item.id == id => {
             contracts.iter().find_map(|contract| match &contract.types {
-                Types(AbsType::Record(row)) => Some(extract_ident(&row)),
-                Types(AbsType::Flat(term)) => {
-                    if let Term::Record(data, ..) | Term::RecRecord(data, ..) = term.as_ref() {
-                        Some(data.fields.keys().cloned().collect())
-                    } else {
-                        None
-                    }
-                }
+                Types(AbsType::Record(row)) => Some(find_fields_from_type(&row, path)),
                 _ => None,
             })
         }
         _ if item.id == id => match item.kind {
             TermKind::Declaration(_, _, ValueState::Known(new_id))
             | TermKind::Usage(UsageState::Resolved(new_id)) => {
-                find_contract_record_fields(&linearization, new_id)
+                find_fields_from_contract(&linearization, new_id, path)
             }
             _ => None,
         },
         _ => None,
     }
 }
 
-/// Extract identifiers from a row type which forms a record.
-/// Ensure that ty is really a row type.
-fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+/// Extract the fields from a given record type.
+fn find_fields_from_type(ty: &Box<Types>, path: &mut Vec<Ident>) -> Vec<Ident> {
+    let current = path.pop();
     ty.iter_as_rows()
-        .filter_map(|item| match item {
-            RowIteratorItem::Row(ident, _) => Some(ident.clone()),
-            RowIteratorItem::Tail(..) => None,
+        .flat_map(|item| match (item, current) {
+            (RowIteratorItem::Row(ident, Some(ty)), Some(current)) if current == *ident => {
+                if let Types(AbsType::Record(ty)) = ty {
+                    find_fields_from_type(ty, path)
+                } else {
+                    find_fields_from_type(&Box::new(ty.clone()), path)
```

Comment:
I'm not entirely sure I'm following the logic here - if `ty` wasn't a record, then either (1) `iter_as_rows` will return nothing or (2) it's an enum, in which case it doesn't have field names. Do we really need to do the recursive call, or can we just give up?

Reply:
Yeah, we don't need it. I was trying to check if it was possible to get a  case where `ty`  could be a `AbsType::Flat(..)`, representing a record contract.

Reply:
Thanks!